### PR TITLE
Csharp Multithreading

### PR DESCRIPTION
**Pull request by Fan1117**

### CODE REVIEW

**Comment by Andrey1994** [comment]
do NOT override existing test we need to keep it - create a new one, do NOT comment out ~20 lines of code its terrible, keep existing code style.

Also lets make it more complicated in terms of logic - request data in both threads simultaneously, add multiple start_stream stop_stream correctly and probably multiple prepare_session release_session too

**Comment by Andrey1994** [comment]
and second thread should request data in a loop like while(keep_alive)

**Comment by Andrey1994** [comment]
also lets move this C# test outside brainflow C# solution and put it in C# test folder(you will need to add a reference to brainflow.dll)

**Comment by Fan1117** [comment]
i will fix this later this week.

**Comment by Andrey1994** [comment]
I think you forgot to add new files to git and updated .travis.yml, also configure your VS to add spaces before brackets its not hard and will keep existing code style

**Comment by Andrey1994** [comment]
now you've copypasted entire c# binding to test folder....
you need to keep C# binding where it is currently located, compile it to C# DLL, create new solution with a reference to this C# dll and use brainflow API in new test

**Comment by Andrey1994** [comment]
take a look at cpp and java tests I dont copypaste entire binding to compile it there are jar

**Comment by Andrey1994** [comment]
Sorry I pressed wrong button I was not going to close it, reopen

**Comment by Andrey1994** [comment]
now it looks much more better, nice! lets fix review comments I will merge it

**Comment by Fan1117** [comment]
I appreciate your help! Thanks a lot! Btw,  what you mean by fixing review comments?

**Comment by Andrey1994** [comment]
comments above which are attached to specific lines of code

**Comment by Andrey1994** [comment]
one more idea for this test, since we now publish nuget package lets test it!
now in deploy_script CI creates nuget package, lets create it during initialization and install this package in test project

File: tests/csharp/multithread/test/get_board_data.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                thread_collector_array[i] = new ThreadCollector(board_shim, args);
+            }
+
+            // prepare and start streaming
+            board_shim.prepare_session();
+            Console.WriteLine("Session is ready");
+
+            board_shim.start_stream(3600);
+            Console.WriteLine("Started");
+            System.Threading.Thread.Sleep(5000);
```
[the remark this answers is not in the document]
Comment (Fan1117) [reply tests/csharp/multithread/test/get_board_data.cs]: I can delete it if you want. Just want to have a little more data (~5s).

File: tests/csharp/multithread/test/ThreadCollector.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        public void collect_data()
+        {
+            while (keep_alive)
+            {
+                System.Threading.Thread.Sleep(sleep_ms);
+                Console.WriteLine("data count: {0}", board_shim.get_board_data_count());
+            }
+            double[,] unprocessed_data = board_shim.get_board_data();
```
[the remark this answers is not in the document]
Comment (Fan1117) [reply tests/csharp/multithread/test/ThreadCollector.cs]: My bad. I was thinking to use get_current_board_data but type the wrong method. I will fix this. May I add signal filtering after this merge?

File: tests/csharp/multithread/test/test.csproj [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    <Optimize>true</Optimize>
+    <DebugType>pdbonly</DebugType>
+    <PlatformTarget>x86</PlatformTarget>
+    <ErrorReport>prompt</ErrorReport>
+    <CodeAnalysisRuleSet>MinimumRecommendedRules.ruleset</CodeAnalysisRuleSet>
+    <Prefer32Bit>true</Prefer32Bit>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="brainflow">
+      <HintPath>..\..\build_single\Release\brainflow.dll</HintPath>
```
[the remark this answers is not in the document]
Comment (Fan1117) [reply tests/csharp/multithread/test/test.csproj]: Got it.

**Pull request closed without merging**